### PR TITLE
RegisterForReflection#classNames does not register full hierarchy of the class

### LINKED ISSUE #21985: RegisterForReflection#classNames does not register full hierarchy of the class
### Description

If we use `io.quarkus.runtime.annotations.RegisterForReflection#classNames` to register classes for reflection it does not register classes up in the hierarchy.

On the other hand, if we use `-H:ReflectionConfigurationFiles`, it does register the full class hierarchy.

### Implementation ideas

I think we should look for the full hierarchy here: https://github.com/quarkusio/quarkus/blob/c82f416fd1ad3ddf8ee40cd40c839e9c884757b3/core/deployment/src/main/java/io/quarkus/deployment/steps/RegisterForReflectionBuildStep.java#L66

And also register super classes.

Are we ok to implement this? I can do it.

### PR DESCRIPTION
**Pull request by igorregis**
fix #21985 

- First time PR
- Didn't find any specific tests examples for these classes, open to suggestions
- Added new javadoc comment for public field created on RegisterForReflection annotation
- Didn't feel the need to add new comments, since the changes haven't added nothing much different from previous code state

### CODE REVIEW

File: core/runtime/src/main/java/io/quarkus/runtime/annotations/RegisterForReflection.java [hunk 1]
```diff
@@ -29,6 +29,13 @@
      */
     boolean ignoreNested() default true;
 
+    /**
+     * Whether the full class hierarchy should be processed or not
+     *
+     * This is intended to be used to register types that are going to be serialized
+     */
+    boolean ignoreHierarchy() default true;
```
Comment (radcortez) [inline core/runtime/src/main/java/io/quarkus/runtime/annotations/RegisterForReflection.java:37]: Wondering if this is useful? I can't remember a use case where you want to ignore the hierarchy, because you need the full hierarchy of the class to use it. If you ignore it then you need to register the missing pieces manually.

File: core/deployment/src/main/java/io/quarkus/deployment/steps/RegisterForReflectionBuildStep.java [hunk 2]
```diff
@@ -1,24 +1,32 @@
...
 import io.quarkus.deployment.builditem.nativeimage.ReflectiveClassBuildItem;
+import io.quarkus.deployment.builditem.nativeimage.ReflectiveHierarchyBuildItem;
 import io.quarkus.runtime.annotations.RegisterForReflection;
 
 public class RegisterForReflectionBuildStep {
 
     private static final Logger log = Logger.getLogger(RegisterForReflectionBuildStep.class);
 
+    @Inject
+    BuildProducer<ReflectiveHierarchyBuildItem> reflectiveClassHierarchy;
```
Comment (radcortez) [inline core/deployment/src/main/java/io/quarkus/deployment/steps/RegisterForReflectionBuildStep.java:28]: This can be part of the `@BuildStep` method.

**Review by radcortez — CHANGES_REQUESTED** [review]
Can we also add a test? I recommend looking into https://github.com/quarkusio/quarkus/blob/57f6a9d9cb288cce2b558d793948b8d0e56ea806/integration-tests/opentelemetry-vertx/src/main/java/io/quarkus/it/opentelemetry/vertx/SpanData.java, which was the trigger for the original issue. Ideally, `io.opentelemetry.sdk.trace.data.SpanData` should be the only class required to be registered.

**Comment by radcortez** [comment]
@igorregis Thank you for your time and the PR. It is very welcomed. I've left a couple of comments to discuss but nothing major and we do need a test for this.

File: core/runtime/src/main/java/io/quarkus/runtime/annotations/RegisterForReflection.java (hunk 1 above)
Comment (igorregis) [reply core/runtime/src/main/java/io/quarkus/runtime/annotations/RegisterForReflection.java:37]: I totally agree. I've put that in order to keep backward compatibility regarding current behavior. It's may be unlikely, but this change can cause cascading software behavior changes, for those who have coded expecting the current behavior.

**Comment by radcortez** [comment]
Strange. If there was an issue with the test, the CI would have reported it. I did try to run it locally and it works fine for me. Can you try to rebuild the entire project from the root folder to make sure you don't have any outdated dependencies? I recommend `mvn install -T 6 -Dquickly -DskipTests=true` from the root folder, and then run the test again.

**Comment by igorregis** [comment]
> Strange. If there was an issue with the test, the CI would have reported it. I did try to run it locally and it works fine for me. Can you try to rebuild the entire project from the root folder to make sure you don't have any outdated dependencies? I recommend `mvn install -T 6 -Dquickly -DskipTests=true` from the root folder, and then run the test again.

You were right, I did a git pull and now everything is ok. Thank you for the tip. I'm gonna work on the tests this weekend.

**Comment by radcortez** [comment]
Great. Thanks!

**Comment by igorregis** [comment]
> Can we also add a test? I recommend looking into https://github.com/quarkusio/quarkus/blob/57f6a9d9cb288cce2b558d793948b8d0e56ea806/integration-tests/opentelemetry-vertx/src/main/java/io/quarkus/it/opentelemetry/vertx/SpanData.java, which was the trigger for the original issue. Ideally, `io.opentelemetry.sdk.trace.data.SpanData` should be the only class required to be registered.

Thank you for the reference for this integration test. It's validating quite well the implementation.  But I've found that the ReflectiveHierarchStep class requires that all class into the hierarchy must be previously  loaded on Jandex index, otherwise it will be skip at this line:

https://github.com/quarkusio/quarkus/blob/57f6a9d9cb288cce2b558d793948b8d0e56ea806/core/deployment/src/main/java/io/quarkus/deployment/steps/ReflectiveHierarchyStep.java#L221-L223

Even as being detected as required with all flags tirned on.

I going to implement that loading at the RegisterForReflectionBuildStep, but that will require a code set very similar to ReflectiveHierarchStep

File: core/runtime/src/main/java/io/quarkus/runtime/annotations/RegisterForReflection.java (hunk 1 above)
Comment (igorregis) [reply core/runtime/src/main/java/io/quarkus/runtime/annotations/RegisterForReflection.java:37]: You need the inherited fields otherwise you will lose information within serialization.

Comment (Sanne) [reply core/runtime/src/main/java/io/quarkus/runtime/annotations/RegisterForReflection.java:37]: Ok but in most cases when we register for reflection we don't do it for Serialization purposes?

Comment (Sanne) [reply core/runtime/src/main/java/io/quarkus/runtime/annotations/RegisterForReflection.java:37]: Let's remember that each reflective registration can potentially trigger a significant overhead, as the compiler can't fold type cardinality individually in each callsite: a reflective type registration is global and can potentially appear anywhere.

So let's try to nudge towards being more selective rather than more inclusive - when it makes sense.

Comment (radcortez) [reply core/runtime/src/main/java/io/quarkus/runtime/annotations/RegisterForReflection.java:37]: Are you saying that you prefer to add a way to control if registration should happen for parent classes?

Comment (Sanne) [reply core/runtime/src/main/java/io/quarkus/runtime/annotations/RegisterForReflection.java:37]: yes I think it's useful and that it should default to `true` for the general case.

Perhaps if the intended use is serialization we should have a different build item to make the intent clearer? Then for the serialization case you could remove this flag and always include parents.

Comment (igorregis) [reply core/runtime/src/main/java/io/quarkus/runtime/annotations/RegisterForReflection.java:37]: > Perhaps if the intended use is serialization we should have a different build item to make the intent clearer? 


@Sanne, do you mean the implementation could be better by creating a BuildStep and a BuildItem specific for that use case and also a annotation?

Something like:
RegisterForSerializationBuildStep
RegisterForSerializationBuildItem
RegisterForSerialization (annotation)

For this case the default would be full hierarchy processing.

Or the RegisterForReflectionBuildStep could be reused for this new BuildItem? If that makes any sense.

Comment (Sanne) [reply core/runtime/src/main/java/io/quarkus/runtime/annotations/RegisterForReflection.java:37]: yes right I'm wondering if that makes sense. However while I'm familiar with reflection, I don't have experience with Serialization in native-image so don't take my word for it: I'm just wondering if that makes sense, based on your comments as it seems one would want slightly different processing & defaults.

This way you could avoid having an attribute for `ignoreHierarchy`: it would always be off for reflection, and always on for serialization to provide better APIs.

Comment (igorregis) [reply core/runtime/src/main/java/io/quarkus/runtime/annotations/RegisterForReflection.java:37]: I can implement that alternative and submit another pr, referencing this one, so we can compare both approaches. 

Is that ok?

**Comment by igorregis** [comment]
> > Can we also add a test? I recommend looking into https://github.com/quarkusio/quarkus/blob/57f6a9d9cb288cce2b558d793948b8d0e56ea806/integration-tests/opentelemetry-vertx/src/main/java/io/quarkus/it/opentelemetry/vertx/SpanData.java, which was the trigger for the original issue. Ideally, `io.opentelemetry.sdk.trace.data.SpanData` should be the only class required to be registered.
> .....
> I'm going to implement that loading at the RegisterForReflectionBuildStep, but that will require a code set very similar to ReflectiveHierarchyStep

@radcortez I guess I have made all needed changes now. Regarding to tests, all I did was to change it to run as the intended behavior as you wrote before:
> Ideally, io.opentelemetry.sdk.trace.data.SpanData should be the only class required to be registered.

Now reading the final code, it feels like it really should be in another buildStep to say at least, because it's clear that the RegisterForReflectionBuildStep seems to have two different responsibilities.

I can, at first, separate the BuildStep logic into another class and create a dedicated new annotation, like RegisterForSerialization for evaluation here. So we can achieve something near to what @Sanne pointed out.

What do you think?

**Comment by radcortez** [comment]
Sure go ahead.

**Comment by igorregis** [comment]
> Sure go ahead.

Hi @radcortez, I appreciate your assistance on this first time endeavor.  I'm sure it's done now. The SpanData have been changed so now it uses the new Annotation and it's passing clearly. 

We have a new BuildStep class and a New explicit annotation that makes clear the serialization intent. I didn't feel the need for a new BuildItem. The ReflectiveHierarchyBuildItem did the trick and it keeps the semantics since we are dealing with a kind of Reflective Hierarchy

If you think it's OK to go ahead, I need some help on cleaning my commits, that seems to be messy :-) Just point me the direction.

> Commits should be atomic and semantic. Please properly squash your pull requests before submitting them. Fixup commits can be used temporarily during the review process but things should be squashed at the end to have meaningful commits. We use merge commits so the GitHub Merge button cannot do that for us. **If you don't know how to do that, just ask in your pull request, we will be happy to help**!

That my case, I don't have a clue of how to do that.

**Comment by radcortez** [comment]
Thanks. Lets see if the CI is happy.

**Comment by igorregis** [comment]
> Thanks. Lets see if the CI is happy.

It seems it didn't like it :-p.

I'm going to work on it next free time that I have.

**Comment by radcortez** [comment]
Thanks!

File: core/deployment/src/main/java/io/quarkus/deployment/steps/RegisterForReflectionBuildStep.java [hunk 3]
```diff
@@ -36,12 +35,6 @@ public void build(BuildProducer<ReflectiveClassBuildItem> reflectiveClass) {
             AnnotationValue classNamesValue = i.value("classNames");
 
             ClassLoader classLoader = Thread.currentThread().getContextClassLoader();
-            if (targetsValue == null && classNamesValue == null) {
-                ClassInfo classInfo = i.target().asClass();
-                registerClass(classLoader, classInfo.name().toString(), methods, fields, ignoreNested, serialization,
-                        reflectiveClass);
-                continue;
-            }
```
Comment (gsmet) [inline core/deployment/src/main/java/io/quarkus/deployment/steps/RegisterForReflectionBuildStep.java:44]: Why this change? When `RegisterForReflection` is on a class without any of these attributes defined, the class itself should be registered for reflection.

File: core/deployment/src/main/java/io/quarkus/deployment/steps/RegisterForSerializationBuildStep.java [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+                for (String className : classNames) {
+                    addClassNameForReflection(reflectiveClassHierarchy, classLoader, processedReflectiveHierarchies,
+                            annotation, methods, fields,
+                            builder, className);
+                }
+            }
+        }
+    }
+
+    private void addClassNameForReflection(BuildProducer<ReflectiveHierarchyBuildItem> reflectiveClassHierarchy,
```
Comment (gsmet) [inline core/deployment/src/main/java/io/quarkus/deployment/steps/RegisterForSerializationBuildStep.java:83]: Should the method be named `ForSerialization`?

**Review by gsmet — CHANGES_REQUESTED** [review]
Added a few comments. As Roberto mentioned, this will need a big squash at the end.

**Review by zakkak — CHANGES_REQUESTED** [review]
@igorregis welcome to Quarkus and thank you for opening this PR :)

The way I understand it, the intention of this PR was to give Quarkus applications the ability to register the whole class hierarchy of a class (or a set of classes) with a single annotation.

IMO this should be implemented through an optional parameter being passed to `RegisterForReflection`, e.g. a boolean called `includeHierarchy` and not in a separate annotation. But even if a separate annotation is dimmed necessary it shouldn't be called `RegisterForSerialization` as such an annotation is expected to do something else.

Interestingly though, `RegisterForSerialization` is an annotation we have been wanting to implement for some time now (see https://github.com/quarkusio/quarkus/issues/20300), so I would suggest keeping it as well (addressing the issues I mention in my review).

So to sum up, I believe that this PR should:
1. Extend `RegisterForReflection` to allow it to register the whole hierarchy
2. Deprecate `RegisterForReflection`'s `serialization` parameter
3. Introduce `RegisterForSerialization` (as it's already there) which should register the class and its hierarchy for reflection including all fields (ideally excluding transient ones) and ideally excluding all methods, except for `writeObject`, `readObject` and `readObjectNoData`, as they are not needed for serialization.

File: core/deployment/src/main/java/io/quarkus/deployment/steps/RegisterForReflectionBuildStep.java (hunk 2 above)
Comment (igorregis) [reply core/deployment/src/main/java/io/quarkus/deployment/steps/RegisterForReflectionBuildStep.java]: This change was addressed on current PR revision

File: core/deployment/src/main/java/io/quarkus/deployment/steps/RegisterForSerializationBuildStep.java (hunk 4 above)
Comment (igorregis) [reply core/deployment/src/main/java/io/quarkus/deployment/steps/RegisterForSerializationBuildStep.java:83]: Would the RegisterForSerializationBuildStep class name be a good string to be used for the source property, as the buildSetp that is generating all this?

Or should it be the actual class name that is being processed?

**Comment by igorregis** [comment]
> Added a few comments. As Roberto mentioned, this will need a big squash at the end.

Actually [here](https://github.com/quarkusio/quarkus/pull/24474#issuecomment-1105648300) I was quoting the quarkus documentation :-) since I surely need some orientation on how to perform the squash. Any link to any documentation or tutorial is suffice.

File: core/deployment/src/main/java/io/quarkus/deployment/steps/RegisterForSerializationBuildStep.java [hunk 5]
```diff
@@ -0,0 +1,216 @@
...
+            AnnotationValue targetsValue = annotation.value("targets");
+            AnnotationValue classNamesValue = annotation.value("classNames");
+
+            ClassLoader classLoader = Thread.currentThread().getContextClassLoader();
+
+            Set<DotName> processedReflectiveHierarchies = new HashSet<DotName>();
+
+            ReflectiveHierarchyBuildItem.Builder builder = new ReflectiveHierarchyBuildItem.Builder();
+            builder.ignoreFieldPredicate(new IgnoreFieldPredicate(fields))
+                    .ignoreMethodPredicate(new IgnoreMethodPredicate(methods));
```
[the remark this answers is not in the document]
Comment (igorregis) [reply core/deployment/src/main/java/io/quarkus/deployment/steps/RegisterForSerializationBuildStep.java:55]: Would the RegisterForSerializationBuildStep class name be a good string to be used for the source property, as the buildSetp that is generating all this?

Or should it be the actual class name that is being processed?

File: core/deployment/src/main/java/io/quarkus/deployment/steps/RegisterForReflectionBuildStep.java (hunk 3 above)
Comment (igorregis) [reply core/deployment/src/main/java/io/quarkus/deployment/steps/RegisterForReflectionBuildStep.java]: You are right it's undue change. Already reverted. Sorry for that.

File: core/deployment/src/main/java/io/quarkus/deployment/steps/RegisterForSerializationBuildStep.java (hunk 4 above)
Comment (igorregis) [reply core/deployment/src/main/java/io/quarkus/deployment/steps/RegisterForSerializationBuildStep.java]: Correct, I've changed all old names now.

File: core/runtime/src/main/java/io/quarkus/runtime/annotations/RegisterForSerialization.java [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+ */
+@Retention(RetentionPolicy.RUNTIME)
+@Target(ElementType.TYPE)
+public @interface RegisterForSerialization {
+
+    /**
+     * If the methods should be registered
+     */
+    boolean methods() default true;
+
```
[the remark this answers is not in the document]
Comment (igorregis) [reply core/runtime/src/main/java/io/quarkus/runtime/annotations/RegisterForSerialization.java:24]: The code is manipulating the methods in order to process the return and parameter types. I need to do some testing to be sure it's not needed. But what you mentioned makes sense

File: integration-tests/opentelemetry-vertx/src/main/java/io/quarkus/it/opentelemetry/vertx/SpanData.java [hunk 7]
```diff
@@ -1,28 +1,7 @@
...
-        "io.opentelemetry.api.trace.ArrayBasedTraceState",
-        "io.opentelemetry.api.trace.AutoValue_ArrayBasedTraceState",
-        "io.opentelemetry.sdk.common.InstrumentationLibraryInfo",
-        "io.opentelemetry.sdk.common.AutoValue_InstrumentationLibraryInfo",
-        "io.opentelemetry.sdk.resources.Resource",
-        "io.opentelemetry.sdk.resources.AutoValue_Resource",
-        "io.opentelemetry.api.common.Attributes",
-        "io.quarkus.opentelemetry.runtime.tracing.DelayedAttributes"
-})
+@RegisterForSerialization(classNames = { "io.opentelemetry.sdk.trace.data.SpanData" })
```
[the remark this answers is not in the document]
Comment (igorregis) [reply integration-tests/opentelemetry-vertx/src/main/java/io/quarkus/it/opentelemetry/vertx/SpanData.java:5]: Good points @zakkak!

Since this integration test aim to test the successful serialization of SpanData class, wouldn't be better to keep the use of the new annotation only? Or am I missing the test goal? (I'm writing this because I haven't gone through the IC results yet)

By the way, with the new annotation, I'm understanding that there is no need for the includeHierarchy attribute into the RegisterForReflection annotation, that's because the original issue was asking for a way to register the full hierarchy for serialization purpose only. 

Actually it was even being implemented by adding this new attribute into RegisterForReflection , but @Sanne helped to understand, [with these comments](https://github.com/quarkusio/quarkus/pull/24474#discussion_r847543455), that a new annotation would be a better way to go. That way, it seems there is no need for the new attribute for RegisterForReflection.

But, what about the serialization attribute from RegisterForReflection annotation? What is it used for today?

File: core/runtime/src/main/java/io/quarkus/runtime/annotations/RegisterForSerialization.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (zakkak) [reply core/runtime/src/main/java/io/quarkus/runtime/annotations/RegisterForSerialization.java:24]: @igorregis can you please clarify to which code you are referring to when saying "the code"?

[the remark this answers is not in the document]
Comment (igorregis) [reply core/runtime/src/main/java/io/quarkus/runtime/annotations/RegisterForSerialization.java:24]: Sorry for that @zakkak, I was referring to [this method](https://github.com/quarkusio/quarkus/pull/24474/files#diff-152ea78045c4169902158c274ce4cd6e4d9a79f0676f71dfa1e6c1dc969300a8R128-R144). Since it's trying to deal only with getters and setters, it's seems that the "method" property could be removed from the annotation, but this method should be maintained so the change can behave the way you commented here:

> 2. Introduce RegisterForSerialization (as it already does) which should register the class and its hierarchy for reflection including all fields (ideally excluding transient ones) and ideally excluding all methods, except for writeObject, readObject and readObjectNoData, as they are not needed for serialization.

[the remark this answers is not in the document]
Comment (zakkak) [reply core/runtime/src/main/java/io/quarkus/runtime/annotations/RegisterForSerialization.java:24]: I am still not convinced that serialization/deserialization requires reflective access to all potential getters and setters. The implementation should (at least in theory) only rely on the `writeObject`, `readObject` and `readObjectNoData` methods (see [here](https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/io/Serializable.html)). Anything else should be explicitly registered by the user IMO. @Sanne WDYT?

[the remark this answers is not in the document]
Comment (Sanne) [reply core/runtime/src/main/java/io/quarkus/runtime/annotations/RegisterForSerialization.java:24]: sorry I don't have experience with Serialization in native. Sounds like adding some tests would be nice?

BTW that's why I was suggesting to keep the builditems separate, one for each purpose: if - as a user of wuch build items - I'm needing *serialization* I shouldn't be using a builditem for *reflection* - if indeed serialization required reflection under the hood, that's an implementation detail.

[the remark this answers is not in the document]
Comment (igorregis) [reply core/runtime/src/main/java/io/quarkus/runtime/annotations/RegisterForSerialization.java:24]: I may be missing something here, but as fas as I undertood it, those methods are optional ([see here](https://docs.oracle.com/en/java/javase/17/docs/specs/serialization/serial-arch.html#the-serializable-interface)) and should be implemented only if the serializable object need something special for its serialization process.

[the remark this answers is not in the document]
Comment (zakkak) [reply core/runtime/src/main/java/io/quarkus/runtime/annotations/RegisterForSerialization.java:24]: Hi @igorregis, you are right. Defining these methods is optional **but** when they are defines they need to be registered for reflection. The reflective access of these methods (when present) happens [here](https://github.com/openjdk/jdk17u-dev/blob/4b1bba808e659dc5ad18d1696bfe564a1df6d3e9/src/java.base/share/classes/java/io/ObjectStreamClass.java#L412-L421)

[the remark this answers is not in the document]
Comment (radcortez) [reply core/runtime/src/main/java/io/quarkus/runtime/annotations/RegisterForSerialization.java:24]: I think we shouldn't mix Serialization with the original intent of the issue, which is, being able to marshall / unmarshall (usually by Jackson or other binding libraries that rely on reflection) a specific object and its components without having to know the entire object structure.

**Review by radcortez — CHANGES_REQUESTED** [review]

File: core/runtime/src/main/java/io/quarkus/runtime/annotations/RegisterForSerialization.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (igorregis) [reply core/runtime/src/main/java/io/quarkus/runtime/annotations/RegisterForSerialization.java:24]: Sorry for the delay, I've been on work trip that took all of my free time. Based on all of that discussion, this implementation should:
a) Enable a class to be correctly registered to be manipulated through reflection **OR**
b) Enable a class to be correctly registered to be serialized **OR**
c) Enable a class to be manipulated through reflection so its state can be retrieved and transferred through some IO process.

I understand that you are pointing on c) option @radcortez.

I have never read the source code from a library like Jackson deep enough to see if they comply with class level serialization or if they introspect the field values through reflection. I suspect it's the later option, based on tests results from CI. But I would need to do some research  in order to confirm that, if you think it's a good idea.

By any case, the problem may be the annotation name that need to be chosen in order to express the final intent.
Based on the three options above we have:
a) @RegisterForReflection
b) @RegisterForSerialization
c) @RegisterForObjectMapping **or** @RegisterForDataBiding **or even** @RegisterForJSONSerialization

[the remark this answers is not in the document]
Comment (radcortez) [reply core/runtime/src/main/java/io/quarkus/runtime/annotations/RegisterForSerialization.java:24]: Yes, I stumbled upon this when doing the JSON marshaling / unmarshalling, but I don't want to make this a specific thing of JSON. We probably need to take one step back and look into the originally reported enhancement.

When registering a class for reflection using `-H:ReflectionConfigurationFiles` and GraalVM configuration style, the entire class hierarchy is added. When using our annotation, only the specified class is registered, and you need to explore all the classes to ensure everything is registered properly which is inconvenient.

The proposed enhancement was to make the register of reflection via the file vs via the annotation more consistent and alike.

I believe that the Serialization discussion is a completely different topic and should not be mixed with the original feature enhancement proposal.

File: integration-tests/opentelemetry-vertx/src/main/java/io/quarkus/it/opentelemetry/vertx/SpanData.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (igorregis) [reply integration-tests/opentelemetry-vertx/src/main/java/io/quarkus/it/opentelemetry/vertx/SpanData.java:5]: So, can we agree to change everything back the original naming (RegisterForReflection), so I'll merge back everything to the existent annotation and buildstep, and treat this change as a behavior fix the way you clarified [here](https://github.com/quarkusio/quarkus/pull/24474#discussion_r901625012) @radcortez.

I think is better to let the creation of a new annotation on the feature request that @zakkak pointed out [here](https://github.com/quarkusio/quarkus/pull/24474#pullrequestreview-964522528).

Can I do that way and submit the new changes for CI and review?

File: core/runtime/src/main/java/io/quarkus/runtime/annotations/RegisterForSerialization.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (geoand) [reply core/runtime/src/main/java/io/quarkus/runtime/annotations/RegisterForSerialization.java:24]: > I believe that the Serialization discussion is a completely different topic and should not be mixed with the original feature enhancement proposal.

I completely agree

File: integration-tests/opentelemetry-vertx/src/main/java/io/quarkus/it/opentelemetry/vertx/SpanData.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (geoand) [reply integration-tests/opentelemetry-vertx/src/main/java/io/quarkus/it/opentelemetry/vertx/SpanData.java:5]: I have not read all the comments in this thread, but when I was looking at the original issue, my idea was to simply add a flag to `@RegisterForReflection` which would control whether or not the entire hierarchy would be registered (the flag would be `false` by default to preserve the existing behavior).

[the remark this answers is not in the document]
Comment (igorregis) [reply integration-tests/opentelemetry-vertx/src/main/java/io/quarkus/it/opentelemetry/vertx/SpanData.java:5]: Great! I'm going to perform the changes and submit it until the end of this week.

[the remark this answers is not in the document]
Comment (igorregis) [reply integration-tests/opentelemetry-vertx/src/main/java/io/quarkus/it/opentelemetry/vertx/SpanData.java]: @radcortez the changes are done as specified

File: core/deployment/src/main/java/io/quarkus/deployment/steps/RegisterForReflectionBuildStep.java [hunk 8]
```diff
@@ -45,49 +64,159 @@ public void build(CombinedIndexBuildItem combinedIndexBuildItem,
...
+            if (methodInfo.parameters().size() > 0 ||
+                    Modifier.isStatic(methodInfo.flags()) ||
+                    methodInfo.returnType().kind() == Kind.VOID || methodInfo.returnType().kind() == Kind.PRIMITIVE) {
+                continue;
+            }
+            registerType(reflectiveClassHierarchy, classLoader, processedReflectiveHierarchies,
+                    methods, builder,
+                    getMethodReturnType(indexView, initialName, classInfo, methodInfo));
+        }
+    }
```
Comment (zakkak) [inline core/deployment/src/main/java/io/quarkus/deployment/steps/RegisterForReflectionBuildStep.java:174]: Non-blocker: I still don't see why serialization/deserialization would use the getters instead of reflection to access the fields. If you could provide an example where getters are accessed/used I would appreciate it :)

File: core/deployment/src/main/java/io/quarkus/deployment/steps/RegisterForReflectionBuildStep.java [hunk 9]
```diff
@@ -45,49 +64,159 @@ public void build(CombinedIndexBuildItem combinedIndexBuildItem,
...
+        if (type.kind() != Kind.PRIMITIVE && !processedReflectiveHierarchies.contains(type.name())) {
+            registerClassDependencies(reflectiveClassHierarchy, classLoader, processedReflectiveHierarchies,
+                    methods, builder, type.name().toString());
+        }
+    }
+
+    private static Type getMethodReturnType(IndexView indexView, DotName initialName, ClassInfo info, MethodInfo method) {
+        Type methodReturnType = method.returnType();
+        if ((method.returnType().kind() == Kind.TYPE_VARIABLE) && (info.typeParameters().size() == 1)) {
+            if (method.returnType().asTypeVariable().identifier().equals(info.typeParameters().get(0).identifier())) {
```
Comment (zakkak) [inline core/deployment/src/main/java/io/quarkus/deployment/steps/RegisterForReflectionBuildStep.java:206]:
```suggestion
        if ((methodReturnType.kind() == Kind.TYPE_VARIABLE) && (info.typeParameters().size() == 1) &&
            methodReturnType.asTypeVariable().identifier().equals(info.typeParameters().get(0).identifier())) {
```

File: core/deployment/src/main/java/io/quarkus/deployment/steps/RegisterForReflectionBuildStep.java [hunk 10]
```diff
@@ -45,49 +64,159 @@ public void build(CombinedIndexBuildItem combinedIndexBuildItem,
...
+                try {
+                    List<Type> types = JandexUtil.resolveTypeParameters(initialName, info.name(), indexView);
+                    if (types.size() == 1) {
```
Comment (zakkak) [inline core/deployment/src/main/java/io/quarkus/deployment/steps/RegisterForReflectionBuildStep.java:209]: Isn't this expected to be always true at this point?

File: core/deployment/src/main/java/io/quarkus/deployment/steps/RegisterForReflectionBuildStep.java [hunk 11]
```diff
@@ -45,49 +64,159 @@ public void build(CombinedIndexBuildItem combinedIndexBuildItem,
...
+                        methodReturnType = types.get(0);
+                    }
+                } catch (IllegalArgumentException ignored) {
+
+                }
```
Comment (zakkak) [inline core/deployment/src/main/java/io/quarkus/deployment/steps/RegisterForReflectionBuildStep.java:214]:
```suggestion
                List<Type> types = JandexUtil.resolveTypeParameters(initialName, info.name(), indexView);
                if (types.size() == 1) {
                    methodReturnType = types.get(0);
                }
```

**Review by zakkak — CHANGES_REQUESTED** [review]
Thanks for the changes @igorregis . Please see a couple of comments, only the one with the try catch is a blocker for me.

**Comment by igorregis** [comment]
> Thanks for the changes @igorregis . Please see a couple of comments, only the one with the try catch is a blocker for me.

Ok, I m on a vacation trip. So I'm going to be able to do this changes in 2 weeks.

File: core/deployment/src/main/java/io/quarkus/deployment/steps/RegisterForReflectionBuildStep.java (hunk 8 above)
Comment (igorregis) [reply core/deployment/src/main/java/io/quarkus/deployment/steps/RegisterForReflectionBuildStep.java:174]: Sure. We are not using the getters per se, but only to guess what field should we register the type from. That way we are registering types returned by public methods, since I'm guessing that its state is needed to be transferred or stored.  

If I stop doing that, the opentelemetry vertex integration test break due to lack of type when you run it on native.

Comment (zakkak) [reply core/deployment/src/main/java/io/quarkus/deployment/steps/RegisterForReflectionBuildStep.java:174]: Why not register the types of all the fields though? Wouldn't that be safer?

**Review by zakkak — APPROVED** [review]
LGTM, thanks for addressing all the blocking issues @igorregis

File: core/deployment/src/main/java/io/quarkus/deployment/steps/RegisterForReflectionBuildStep.java (hunk 8 above)
Comment (igorregis) [reply core/deployment/src/main/java/io/quarkus/deployment/steps/RegisterForReflectionBuildStep.java:174]: That crossed my mind, but I've noticed that many types returned on opentelemetry vertex integration test are just interfaces, so they have no fields to be registered and we have no access (the developer may not know and not asking them to be registered) to the real objects behind those interfaces.

**Comment by igorregis** [comment]
My fork was too behind so I discarded all changes on my main. I had no idea it would close the PR, sorry for that. I still have all history on my fork's feature branch.
What's the preferred option now?

**Comment by zakkak** [comment]
> My fork was too behind so I discarded all changes on my main.

FYI: In such cases `git rebase` often does the trick without any hustle if there are not conflicts.

> I had no idea it would close the PR, sorry for that. 

That's OK.

> I still have all history on my fork's feature branch. What's the preferred option now? Should I simply throw away all change history (IMHO it doesn't seems to be important anyway) and do a fresh PR with only one commit?

Yes I think opening a new PR that will be referencing this one is OK.

> @[igorregis](https://github.com/igorregis) igorregis merged commit [b5618c6](https://github.com/quarkusio/quarkus/commit/b5618c6be0b37e094cde89c2c6b2d5f8a431bbed) into quarkusio:main [2 days ago](https://github.com/quarkusio/quarkus/pull/24474#event-7228182211)

That got me really confused :)
@gsmet you might want to keep this in mind. I think it will make this PR's title end in the changelog despite no actual code going in.

**Comment by humcqc** [comment]
Hi @igorregis , thanks for your quick reply,
For my first point [here](https://github.com/quarkusio/quarkus/pull/36594/files#diff-84d7b0556d6497f56447c5da575d426247e8cd4ff64132e5cbe4de8b76229ff5) is my example : when I registered PersonDTO for reflection I was expecting the members of this class to be registered by default. But it was not if I don't add the registerFullHierarchy = true.

For the second point:
[Here](https://github.com/quarkusio/quarkus/pull/36594/files#diff-fe62252ed9d041505b37ef01f152c423413ac1a531aef1134e3b2955d182ddcc) is my annotation.
And [here](https://github.com/quarkusio/quarkus/pull/36594/files#diff-84d7b0556d6497f56447c5da575d426247e8cd4ff64132e5cbe4de8b76229ff5R24) where I use it.
My first thought was that it will register the class annotated by my annotation, but not.
Do you know how I can enhance the registration processor to handle this case if we think it's a good idea ?
